### ISSUE TITLE
[CT-1386] [Bug] `dbt docs generate` with `--defer`

### ISSUE THREAD

**Issue by seub** [dbt-labs/dbt-core#6124]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

It's great that dbt 1.3 makes it possible to use `dbt docs generate` with `--defer`, but it appears to be not quite working, unless I'm not understanding something.

When running `dbt docs generate (--select state:modified+) --defer --state /path/to/manifest` I'm getting an error `Object X does not exist or not authorized.` The cause for this error is that X indeed does not exist, so the `ref` referring to it should have been resolved using the manifest.

In my situation, the faulty `ref` actually happens in a macro, perhaps that's why it wasn't resolved properly? What is strange is that this doesn't seem to be a problem for `dbt run`.

### Expected Behavior

When using `dbt docs generate` with `--defer`, all `ref` calls should be resolved as described in the [dbt doc](https://docs.getdbt.com/reference/node-selection/defer).

By the way, it's unclear to me whether I'm supposed to use a `--select` when doing `dbt docs generate` with `--defer`. I would assume so, since it's supposed to play a role in resolving  the `ref`s according to the doc, but it doesn't seem to make a difference.

### Steps To Reproduce

1. Create a macro that `ref`s some model X.
2. Run `dbt docs generate` with `--defer` (and possibly other flags).
3. Observe that the `ref` in the macro is not resolved properly and produces an error.

### Environment

```markdown
- OS:
- Python:
- dbt:
```

### Which database adapter are you using with dbt?

snowflake

**Comment by jtcohen6** [comment]
Thanks for opening @seub! After a little bit of digging, I think I figured out what's going on here.

Even though https://github.com/dbt-labs/dbt-core/pull/4514 did succeed in moving the `defer_to_manifest` method into the `compile` task (from the `run` task), and adding `--defer` as a supported argument to `docs generate`, the call to `defer_to_manifest` is still missing from the `before_run` steps for `compile` / `docs generate`.

https://github.com/dbt-labs/dbt-core/blob/8145eed603266951ce35858f7eef3836012090bd/core/dbt/task/runnable.py#L420-L422

Needs to become:

```python
    def before_run(self, adapter, selected_uids: AbstractSet[str]):
        with adapter.connection_named("master"):
            self.populate_adapter_cache(adapter)
            self.defer_to_manifest(adapter, selected_uids)
```
And then it works!

Marking this one as a `good_first_issue`, if you're up to make the contribution :) Might be a teensy bit trickier to add an automated test for this, though I'm hopeful that it could be a simple extension of the test added in #4514 ([`test_docs_generate_defer`](https://github.com/dbt-labs/dbt-core/blob/main/tests/functional/artifacts/test_docs_generate_defer.py)).

**Comment by seub** [comment]
Thank you very much @jtcohen6 . I appreciate the encouragement to contribute.

I went through the documentation to contribute and followed all the steps. I just created a [new PR ](https://github.com/dbt-labs/dbt-core/pull/4514) with the change that you suggested. 

However that's the only change I made (besides running `changie`), I did not add any new tests. Note: I was able to see that `make test` succeeds, however `make integration` failed, but it already failed before I made any changes.

**Comment by mivanicova** [comment]
Hi, I created a new pull request #6488 with the same fix as done by @seub in #6188, but I extended integration tests to verify whether catalog is generated from the deferred manifest. I hope that this helps to resolve this issue. Please let me know if I can do anything else/differently.
